File: 1week/근우/3_map-of-highest-peak.py
from collections import deque
from typing import List


class Solution:
    def highestPeak(self, isWater: List[List[int]]) -> List[List[int]]:
        NOT_VISIT = -1
        M, N = len(isWater), len(isWater[0])
        map = [[NOT_VISIT for i in range(N)] for j in range(M)]
        que = deque([])
        for i in range(M):
            for j in range(N):
                if isWater[i][j] == 1:
                    map[i][j] = 0
                    que.append([i, j])

        dy, dx = [0, 0, -1, 1], [1, -1, 0, 0]
        while que:
            cur_pos = que.popleft()
            for i in range(4):
                next_y, next_x = cur_pos[0] + dy[i], cur_pos[1] + dx[i]
                if (
                    0 <= next_y < M
                    and 0 <= next_x < N
                    and map[next_y][next_x] == NOT_VISIT
                ):
                    map[next_y][next_x] = map[cur_pos[0]][cur_pos[1]] + 1
                    que.append([next_y, next_x])
        return map

File: 1week/남규/3. Map of Highest Peak.cpp
//https://leetcode.com/problems/map-of-highest-peak/

class Solution {
public:
    vector<vector<int>> highestPeak(vector<vector<int>>& isWater) {
        int N = isWater.size();
        int M = isWater[0].size();
        vector<vector<int>> ans(N, vector<int>(M, -1));
        
        queue<vector<int>> que;
        
        for(int i = 0; i < N; i++){
            for(int j = 0; j < M; j++){
                if(isWater[i][j]){
                    que.push({i, j});
                    ans[i][j] = 0;
                }
            }
        }
        
        int ys[4] = {1, 0, -1, 0};
        int xs[4] = {0, 1, 0, -1};
        
        if(que.empty()){
            ans.resize(N, vector<int>(M, 0));
            return ans;
        }
        
        while(!que.empty()){
            int y = que.front()[0];
            int x = que.front()[1];
            que.pop();
            
            for(int i = 0; i < 4; i++){
                int nextY = y + ys[i];
                int nextX = x + xs[i];
                
                if(nextY < 0 || nextX < 0 || nextY >= N || nextX >= M){
                    continue;
                }
                
                if(ans[nextY][nextX] != -1){
                    continue;
                }
                ans[nextY][nextX] = ans[y][x] + 1;
                que.push({nextY, nextX});
            }
        }
        return ans;
    }
};
